File: PHP/One/AstTransformer.php
<?php

namespace One\AstTransformer;

use One\Ast\AstTypes\IHasTypeArguments;
use One\Ast\AstTypes\ClassType;
use One\Ast\AstTypes\InterfaceType;
use One\Ast\AstTypes\UnresolvedType;
use One\Ast\AstTypes\LambdaType;
use One\Ast\Expressions\Identifier;
use One\Ast\Expressions\BinaryExpression;
use One\Ast\Expressions\ConditionalExpression;
use One\Ast\Expressions\NewExpression;
use One\Ast\Expressions\TemplateString;
use One\Ast\Expressions\ParenthesizedExpression;
use One\Ast\Expressions\UnaryExpression;
use One\Ast\Expressions\PropertyAccessExpression;
use One\Ast\Expressions\ElementAccessExpression;
use One\Ast\Expressions\ArrayLiteral;
use One\Ast\Expressions\MapLiteral;
use One\Ast\Expressions\Expression;
use One\Ast\Expressions\CastExpression;
use One\Ast\Expressions\UnresolvedCallExpression;
use One\Ast\Expressions\InstanceOfExpression;
use One\Ast\Expressions\AwaitExpression;
use One\Ast\Expressions\StringLiteral;
use One\Ast\Expressions\NumericLiteral;
use One\Ast\Expressions\NullLiteral;
use One\Ast\Expressions\RegexLiteral;
use One\Ast\Expressions\BooleanLiteral;
use One\Ast\Expressions\StaticMethodCallExpression;
use One\Ast\Expressions\InstanceMethodCallExpression;
use One\Ast\Expressions\UnresolvedNewExpression;
use One\Ast\Expressions\NullCoalesceExpression;
use One\Ast\Expressions\UnresolvedMethodCallExpression;
use One\Ast\Expressions\GlobalFunctionCallExpression;
use One\Ast\Expressions\LambdaCallExpression;
use One\Ast\Statements\ReturnStatement;
use One\Ast\Statements\ExpressionStatement;
use One\Ast\Statements\IfStatement;
use One\Ast\Statements\ThrowStatement;
use One\Ast\Statements\VariableDeclaration;
use One\Ast\Statements\WhileStatement;
use One\Ast\Statements\ForStatement;
use One\Ast\Statements\ForeachStatement;
use One\Ast\Statements\Statement;
use One\Ast\Statements\UnsetStatement;
use One\Ast\Statements\BreakStatement;
use One\Ast\Statements\ContinueStatement;
use One\Ast\Statements\DoStatement;
use One\Ast\Statements\TryStatement;
use One\Ast\Statements\Block;
use One\Ast\Types\Method;
use One\Ast\Types\Constructor;
use One\Ast\Types\Field;
use One\Ast\Types\Property;
use One\Ast\Types\Interface_;
use One\Ast\Types\Class_;
use One\Ast\Types\Enum;
use One\Ast\Types\EnumMember;
use One\Ast\Types\SourceFile;
use One\Ast\Types\IVariable;
use One\Ast\Types\IVariableWithInitializer;
use One\Ast\Types\MethodParameter;
use One\Ast\Types\Lambda;
use One\Ast\Types\IMethodBase;
use One\Ast\Types\Package;
use One\Ast\Types\GlobalFunction;
use One\Ast\Types\IInterface;
use One\Ast\Types\IAstNode;
use One\Ast\Types\IHasAttributesAndTrivia;
use One\Ast\Types\Import;
use One\Ast\References\ClassReference;
use One\Ast\References\EnumReference;
use One\Ast\References\ThisReference;
use One\Ast\References\MethodParameterReference;
use One\Ast\References\VariableDeclarationReference;
use One\Ast\References\ForVariableReference;
use One\Ast\References\ForeachVariableReference;
use One\Ast\References\SuperReference;
use One\Ast\References\InstanceFieldReference;
use One\Ast\References\InstancePropertyReference;
use One\Ast\References\StaticPropertyReference;
use One\Ast\References\StaticFieldReference;
use One\Ast\References\CatchVariableReference;
use One\Ast\References\GlobalFunctionReference;
use One\Ast\References\EnumMemberReference;
use One\Ast\References\StaticThisReference;
use One\Ast\References\VariableReference;
use One\ErrorManager\ErrorManager;
use One\ITransformer\ITransformer;
use One\Ast\Interfaces\IType;

class AstTransformer implements ITransformer {
    public $errorMan;
    public $currentFile;
    public $currentInterface;
    public $currentMethod;
    public $currentClosure;
    public $currentStatement;
    public $name;
    
    function __construct($name) {
        $this->name = $name;
        $this->errorMan = new ErrorManager();
        $this->currentFile = null;
        $this->currentInterface = null;
        $this->currentMethod = null;
        $this->currentClosure = null;
        $this->currentStatement = null;
    }
    
    protected function visitAttributesAndTrivia($node) {
        
    }
    
    protected function visitType($type) {
        if ($type instanceof ClassType || $type instanceof InterfaceType || $type instanceof UnresolvedType) {
            $type2 = $type;
            $type2->typeArguments = array_map(function ($x) { return $this->visitType($x); }, $type2->typeArguments);
        }
        else if ($type instanceof LambdaType) {
            foreach ($type->parameters as $mp)
                $this->visitMethodParameter($mp);
            $type->returnType = $this->visitType($type->returnType);
        }
        return $type;
    }
    
    protected function visitIdentifier($id) {
        return $id;
    }
    
    protected function visitVariable($variable) {
        if ($variable->type !== null)
            $variable->type = $this->visitType($variable->type);
        return $variable;
    }
    
    protected function visitVariableWithInitializer($variable) {
        $this->visitVariable($variable);
        if ($variable->initializer !== null)
            $variable->initializer = $this->visitExpression($variable->initializer);
        return $variable;
    }
    
    protected function visitVariableDeclaration($stmt) {
        $this->visitVariableWithInitializer($stmt);
        return $stmt;
    }
    
    protected function visitUnknownStatement($stmt) {
        $this->errorMan->throw("Unknown statement type");
        return $stmt;
    }
    
    protected function visitStatement($stmt) {
        $this->currentStatement = $stmt;
        $this->visitAttributesAndTrivia($stmt);
        if ($stmt instanceof ReturnStatement) {
            if ($stmt->expression !== null)
                $stmt->expression = $this->visitExpression($stmt->expression);
        }
        else if ($stmt instanceof ExpressionStatement)
            $stmt->expression = $this->visitExpression($stmt->expression);
        else if ($stmt instanceof IfStatement) {
            $stmt->condition = $this->visitExpression($stmt->condition);
            $stmt->then = $this->visitBlock($stmt->then);
            if ($stmt->else_ !== null)
                $stmt->else_ = $this->visitBlock($stmt->else_);
        }
        else if ($stmt instanceof ThrowStatement)
            $stmt->expression = $this->visitExpression($stmt->expression);
        else if ($stmt instanceof VariableDeclaration)
            return $this->visitVariableDeclaration($stmt);
        else if ($stmt instanceof WhileStatement) {
            $stmt->condition = $this->visitExpression($stmt->condition);
            $stmt->body = $this->visitBlock($stmt->body);
        }
        else if ($stmt instanceof DoStatement) {
            $stmt->condition = $this->visitExpression($stmt->condition);
            $stmt->body = $this->visitBlock($stmt->body);
        }
        else if ($stmt instanceof ForStatement) {
            if ($stmt->itemVar !== null)
                $this->visitVariableWithInitializer($stmt->itemVar);
            $stmt->condition = $this->visitExpression($stmt->condition);
            $stmt->incrementor = $this->visitExpression($stmt->incrementor);
            $stmt->body = $this->visitBlock($stmt->body);
        }
        else if ($stmt instanceof ForeachStatement) {
            $this->visitVariable($stmt->itemVar);
            $stmt->items = $this->visitExpression($stmt->items);
            $stmt->body = $this->visitBlock($stmt->body);
        }
        else if ($stmt instanceof TryStatement) {
            $stmt->tryBody = $this->visitBlock($stmt->tryBody);
            if ($stmt->catchBody !== null) {
                $this->visitVariable($stmt->catchVar);
                $stmt->catchBody = $this->visitBlock($stmt->catchBody);
            }
            if ($stmt->finallyBody !== null)
                $stmt->finallyBody = $this->visitBlock($stmt->finallyBody);
        }
        else if ($stmt instanceof BreakStatement) { }
        else if ($stmt instanceof UnsetStatement)
            $stmt->expression = $this->visitExpression($stmt->expression);
        else if ($stmt instanceof ContinueStatement) { }
        else
            return $this->visitUnknownStatement($stmt);
        return $stmt;
    }
    
    protected function visitBlock($block) {
        $block->statements = array_map(function ($x) { return $this->visitStatement($x); }, $block->statements);
        return $block;
    }
    
    protected function visitTemplateString($expr) {
        for ($i = 0; $i < count($expr->parts); $i++) {
            $part = $expr->parts[$i];
            if (!$part->isLiteral)
                $part->expression = $this->visitExpression($part->expression);
        }
        return $expr;
    }
    
    protected function visitUnknownExpression($expr) {
        $this->errorMan->throw("Unknown expression type");
        return $expr;
    }
    
    protected function visitLambda($lambda) {
        $prevClosure = $this->currentClosure;
        $this->currentClosure = $lambda;
        $this->visitMethodBase($lambda);
        $this->currentClosure = $prevClosure;
        return $lambda;
    }
    
    protected function visitVariableReference($varRef) {
        return $varRef;
    }
    
    protected function visitExpression($expr) {
        if ($expr instanceof BinaryExpression) {
            $expr->left = $this->visitExpression($expr->left);
            $expr->right = $this->visitExpression($expr->right);
        }
        else if ($expr instanceof NullCoalesceExpression) {
            $expr->defaultExpr = $this->visitExpression($expr->defaultExpr);
            $expr->exprIfNull = $this->visitExpression($expr->exprIfNull);
        }
        else if ($expr instanceof UnresolvedCallExpression) {
            $expr->func = $this->visitExpression($expr->func);
            $expr->typeArgs = array_map(function ($x) { return $this->visitType($x); }, $expr->typeArgs);
            $expr->args = array_map(function ($x) { return $this->visitExpression($x); }, $expr->args);
        }
        else if ($expr instanceof UnresolvedMethodCallExpression) {
            $expr->object = $this->visitExpression($expr->object);
            $expr->typeArgs = array_map(function ($x) { return $this->visitType($x); }, $expr->typeArgs);
            $expr->args = array_map(function ($x) { return $this->visitExpression($x); }, $expr->args);
        }
        else if ($expr instanceof ConditionalExpression) {
            $expr->condition = $this->visitExpression($expr->condition);
            $expr->whenTrue = $this->visitExpression($expr->whenTrue);
            $expr->whenFalse = $this->visitExpression($expr->whenFalse);
        }
        else if ($expr instanceof Identifier)
            return $this->visitIdentifier($expr);
        else if ($expr instanceof UnresolvedNewExpression) {
            $this->visitType($expr->cls);
            $expr->args = array_map(function ($x) { return $this->visitExpression($x); }, $expr->args);
        }
        else if ($expr instanceof NewExpression) {
            $this->visitType($expr->cls);
            $expr->args = array_map(function ($x) { return $this->visitExpression($x); }, $expr->args);
        }
        else if ($expr instanceof TemplateString)
            return $this->visitTemplateString($expr);
        else if ($expr instanceof ParenthesizedExpression)
            $expr->expression = $this->visitExpression($expr->expression);
        else if ($expr instanceof UnaryExpression)
            $expr->operand = $this->visitExpression($expr->operand);
        else if ($expr instanceof PropertyAccessExpression)
            $expr->object = $this->visitExpression($expr->object);
        else if ($expr instanceof ElementAccessExpression) {
            $expr->object = $this->visitExpression($expr->object);
            $expr->elementExpr = $this->visitExpression($expr->elementExpr);
        }
        else if ($expr instanceof ArrayLiteral)
            $expr->items = array_map(function ($x) { return $this->visitExpression($x); }, $expr->items);
        else if ($expr instanceof MapLiteral)
            foreach ($expr->items as $item)
                $item->value = $this->visitExpression($item->value);
        else if ($expr instanceof StringLiteral) { }
        else if ($expr instanceof BooleanLiteral) { }
        else if ($expr instanceof NumericLiteral) { }
        else if ($expr instanceof NullLiteral) { }
        else if ($expr instanceof RegexLiteral) { }
        else if ($expr instanceof CastExpression) {
            $expr->newType = $this->visitType($expr->newType);
            $expr->expression = $this->visitExpression($expr->expression);
        }
        else if ($expr instanceof InstanceOfExpression) {
            $expr->expr = $this->visitExpression($expr->expr);
            $expr->checkType = $this->visitType($expr->checkType);
        }
        else if ($expr instanceof AwaitExpression)
            $expr->expr = $this->visitExpression($expr->expr);
        else if ($expr instanceof Lambda)
            return $this->visitLambda($expr);
        else if ($expr instanceof ClassReference) { }
        else if ($expr instanceof EnumReference) { }
        else if ($expr instanceof ThisReference) { }
        else if ($expr instanceof StaticThisReference) { }
        else if ($expr instanceof MethodParameterReference)
            return $this->visitVariableReference($expr);
        else if ($expr instanceof VariableDeclarationReference)
            return $this->visitVariableReference($expr);
        else if ($expr instanceof ForVariableReference)
            return $this->visitVariableReference($expr);
        else if ($expr instanceof ForeachVariableReference)
            return $this->visitVariableReference($expr);
        else if ($expr instanceof CatchVariableReference)
            return $this->visitVariableReference($expr);
        else if ($expr instanceof GlobalFunctionReference) { }
        else if ($expr instanceof SuperReference) { }
        else if ($expr instanceof InstanceFieldReference) {
            $expr->object = $this->visitExpression($expr->object);
            return $this->visitVariableReference($expr);
        }
        else if ($expr instanceof InstancePropertyReference) {
            $expr->object = $this->visitExpression($expr->object);
            return $this->visitVariableReference($expr);
        }
        else if ($expr instanceof StaticFieldReference)
            return $this->visitVariableReference($expr);
        else if ($expr instanceof StaticPropertyReference)
            return $this->visitVariableReference($expr);
        else if ($expr instanceof EnumMemberReference) { }
        else if ($expr instanceof StaticMethodCallExpression) {
            $expr->typeArgs = array_map(function ($x) { return $this->visitType($x); }, $expr->typeArgs);
            $expr->args = array_map(function ($x) { return $this->visitExpression($x); }, $expr->args);
        }
        else if ($expr instanceof GlobalFunctionCallExpression)
            $expr->args = array_map(function ($x) { return $this->visitExpression($x); }, $expr->args);
        else if ($expr instanceof InstanceMethodCallExpression) {
            $expr->object = $this->visitExpression($expr->object);
            $expr->typeArgs = array_map(function ($x) { return $this->visitType($x); }, $expr->typeArgs);
            $expr->args = array_map(function ($x) { return $this->visitExpression($x); }, $expr->args);
        }
        else if ($expr instanceof LambdaCallExpression)
            $expr->args = array_map(function ($x) { return $this->visitExpression($x); }, $expr->args);
        else
            return $this->visitUnknownExpression($expr);
        return $expr;
    }
    
    protected function visitMethodParameter($methodParameter) {
        $this->visitAttributesAndTrivia($methodParameter);
        $this->visitVariableWithInitializer($methodParameter);
    }
    
    protected function visitMethodBase($method) {
        foreach ($method->parameters as $item)
            $this->visitMethodParameter($item);
        
        if ($method->body !== null)
            $method->body = $this->visitBlock($method->body);
    }
    
    protected function visitMethod($method) {
        $this->currentMethod = $method;
        $this->currentClosure = $method;
        $this->visitAttributesAndTrivia($method);
        $this->visitMethodBase($method);
        $method->returns = $this->visitType($method->returns);
        $this->currentClosure = null;
        $this->currentMethod = null;
    }
    
    protected function visitGlobalFunction($func) {
        $this->visitMethodBase($func);
        $func->returns = $this->visitType($func->returns);
    }
    
    protected function visitConstructor($constructor) {
        $this->currentMethod = $constructor;
        $this->currentClosure = $constructor;
        $this->visitAttributesAndTrivia($constructor);
        $this->visitMethodBase($constructor);
        $this->currentClosure = null;
        $this->currentMethod = null;
    }
    
    protected function visitField($field) {
        $this->visitAttributesAndTrivia($field);
        $this->visitVariableWithInitializer($field);
    }
    
    protected function visitProperty($prop) {
        $this->visitAttributesAndTrivia($prop);
        $this->visitVariable($prop);
        if ($prop->getter !== null)
            $prop->getter = $this->visitBlock($prop->getter);
        if ($prop->setter !== null)
            $prop->setter = $this->visitBlock($prop->setter);
    }
    
    protected function visitInterface($intf) {
        $this->currentInterface = $intf;
        $this->visitAttributesAndTrivia($intf);
        $intf->baseInterfaces = array_map(function ($x) { return $this->visitType($x); }, $intf->baseInterfaces);
        foreach ($intf->fields as $field)
            $this->visitField($field);
        foreach ($intf->methods as $method)
            $this->visitMethod($method);
        $this->currentInterface = null;
    }
    
    protected function visitClass($cls) {
        $this->currentInterface = $cls;
        $this->visitAttributesAndTrivia($cls);
        if ($cls->constructor_ !== null)
            $this->visitConstructor($cls->constructor_);
        
        $cls->baseClass = $this->visitType($cls->baseClass);
        $cls->baseInterfaces = array_map(function ($x) { return $this->visitType($x); }, $cls->baseInterfaces);
        foreach ($cls->fields as $field)
            $this->visitField($field);
        foreach ($cls->properties as $prop)
            $this->visitProperty($prop);
        foreach ($cls->methods as $method)
            $this->visitMethod($method);
        $this->currentInterface = null;
    }
    
    protected function visitEnum($enum_) {
        $this->visitAttributesAndTrivia($enum_);
        foreach ($enum_->values as $value)
            $this->visitEnumMember($value);
    }
    
    protected function visitEnumMember($enumMember) {
        
    }
    
    protected function visitImport($imp) {
        $this->visitAttributesAndTrivia($imp);
    }
    
    function visitFile($sourceFile) {
        $this->errorMan->resetContext($this);
        $this->currentFile = $sourceFile;
        foreach ($sourceFile->imports as $imp)
            $this->visitImport($imp);
        foreach ($sourceFile->enums as $enum_)
            $this->visitEnum($enum_);
        foreach ($sourceFile->interfaces as $intf)
            $this->visitInterface($intf);
        foreach ($sourceFile->classes as $cls)
            $this->visitClass($cls);
        foreach ($sourceFile->funcs as $func)
            $this->visitGlobalFunction($func);
        $sourceFile->mainBlock = $this->visitBlock($sourceFile->mainBlock);
        $this->currentFile = null;
    }
    
    function visitFiles($files) {
        foreach ($files as $file)
            $this->visitFile($file);
    }
    
    function visitPackage($pkg) {
        $this->visitFiles(array_values($pkg->files));
    }
}

File: CSharp/One/AstTransformer.cs
using One.Ast;
using One;
using System.Linq;

namespace One
{
    public class AstTransformer : ITransformer
    {
        public ErrorManager errorMan;
        public SourceFile currentFile;
        public IInterface currentInterface;
        public IMethodBase currentMethod;
        public IMethodBase currentClosure;
        public Statement currentStatement;
        public string name { get; set; }
        
        public AstTransformer(string name)
        {
            this.name = name;
            this.errorMan = new ErrorManager();
            this.currentFile = null;
            this.currentInterface = null;
            this.currentMethod = null;
            this.currentClosure = null;
            this.currentStatement = null;
        }
        
        protected virtual void visitAttributesAndTrivia(IHasAttributesAndTrivia node)
        {
            
        }
        
        protected virtual IType visitType(IType type)
        {
            if (type is ClassType classType || type is InterfaceType || type is UnresolvedType) {
                var type2 = ((IHasTypeArguments)type);
                type2.typeArguments = type2.typeArguments.map(x => this.visitType(x));
            }
            else if (type is LambdaType lambdType) {
                foreach (var mp in lambdType.parameters)
                    this.visitMethodParameter(mp);
                lambdType.returnType = this.visitType(lambdType.returnType);
            }
            return type;
        }
        
        protected virtual Expression visitIdentifier(Identifier id)
        {
            return id;
        }
        
        protected virtual IVariable visitVariable(IVariable variable)
        {
            if (variable.type != null)
                variable.type = this.visitType(variable.type);
            return variable;
        }
        
        protected virtual IVariableWithInitializer visitVariableWithInitializer(IVariableWithInitializer variable)
        {
            this.visitVariable(variable);
            if (variable.initializer != null)
                variable.initializer = this.visitExpression(variable.initializer);
            return variable;
        }
        
        protected virtual VariableDeclaration visitVariableDeclaration(VariableDeclaration stmt)
        {
            this.visitVariableWithInitializer(stmt);
            return stmt;
        }
        
        protected Statement visitUnknownStatement(Statement stmt)
        {
            this.errorMan.throw_($"Unknown statement type");
            return stmt;
        }
        
        protected virtual Statement visitStatement(Statement stmt)
        {
            this.currentStatement = stmt;
            this.visitAttributesAndTrivia(stmt);
            if (stmt is ReturnStatement retStat) {
                if (retStat.expression != null)
                    retStat.expression = this.visitExpression(retStat.expression);
            }
            else if (stmt is ExpressionStatement exprStat)
                exprStat.expression = this.visitExpression(exprStat.expression);
            else if (stmt is IfStatement ifStat) {
                ifStat.condition = this.visitExpression(ifStat.condition);
                ifStat.then = this.visitBlock(ifStat.then);
                if (ifStat.else_ != null)
                    ifStat.else_ = this.visitBlock(ifStat.else_);
            }
            else if (stmt is ThrowStatement throwStat)
                throwStat.expression = this.visitExpression(throwStat.expression);
            else if (stmt is VariableDeclaration varDecl)
                return this.visitVariableDeclaration(varDecl);
            else if (stmt is WhileStatement whileStat) {
                whileStat.condition = this.visitExpression(whileStat.condition);
                whileStat.body = this.visitBlock(whileStat.body);
            }
            else if (stmt is DoStatement doStat) {
                doStat.condition = this.visitExpression(doStat.condition);
                doStat.body = this.visitBlock(doStat.body);
            }
            else if (stmt is ForStatement forStat) {
                if (forStat.itemVar != null)
                    this.visitVariableWithInitializer(forStat.itemVar);
                forStat.condition = this.visitExpression(forStat.condition);
                forStat.incrementor = this.visitExpression(forStat.incrementor);
                forStat.body = this.visitBlock(forStat.body);
            }
            else if (stmt is ForeachStatement forStat2) {
                this.visitVariable(forStat2.itemVar);
                forStat2.items = this.visitExpression(forStat2.items);
                forStat2.body = this.visitBlock(forStat2.body);
            }
            else if (stmt is TryStatement tryStat) {
                tryStat.tryBody = this.visitBlock(tryStat.tryBody);
                if (tryStat.catchBody != null) {
                    this.visitVariable(tryStat.catchVar);
                    tryStat.catchBody = this.visitBlock(tryStat.catchBody);
                }
                if (tryStat.finallyBody != null)
                    tryStat.finallyBody = this.visitBlock(tryStat.finallyBody);
            }
            else if (stmt is BreakStatement) { }
            else if (stmt is UnsetStatement unsetStat)
                unsetStat.expression = this.visitExpression(unsetStat.expression);
            else if (stmt is ContinueStatement) { }
            else
                return this.visitUnknownStatement(stmt);
            return stmt;
        }
        
        protected virtual Block visitBlock(Block block)
        {
            block.statements = block.statements.map(x => this.visitStatement(x)).ToList();
            return block;
        }
        
        protected TemplateString visitTemplateString(TemplateString expr)
        {
            for (int i = 0; i < expr.parts.length(); i++) {
                var part = expr.parts.get(i);
                if (!part.isLiteral)
                    part.expression = this.visitExpression(part.expression);
            }
            return expr;
        }
        
        protected Expression visitUnknownExpression(Expression expr)
        {
            this.errorMan.throw_($"Unknown expression type");
            return expr;
        }
        
        protected virtual Lambda visitLambda(Lambda lambda)
        {
            var prevClosure = this.currentClosure;
            this.currentClosure = lambda;
            this.visitMethodBase(lambda);
            this.currentClosure = prevClosure;
            return lambda;
        }
        
        protected virtual VariableReference visitVariableReference(VariableReference varRef)
        {
            return varRef;
        }
        
        protected virtual Expression visitExpression(Expression expr)
        {
            if (expr is BinaryExpression binExpr) {
                binExpr.left = this.visitExpression(binExpr.left);
                binExpr.right = this.visitExpression(binExpr.right);
            }
            else if (expr is NullCoalesceExpression nullCoalExpr) {
                nullCoalExpr.defaultExpr = this.visitExpression(nullCoalExpr.defaultExpr);
                nullCoalExpr.exprIfNull = this.visitExpression(nullCoalExpr.exprIfNull);
            }
            else if (expr is UnresolvedCallExpression unrCallExpr) {
                unrCallExpr.func = this.visitExpression(unrCallExpr.func);
                unrCallExpr.typeArgs = unrCallExpr.typeArgs.map(x => this.visitType(x));
                unrCallExpr.args = unrCallExpr.args.map(x => this.visitExpression(x));
            }
            else if (expr is UnresolvedMethodCallExpression unrMethCallExpr) {
                unrMethCallExpr.object_ = this.visitExpression(unrMethCallExpr.object_);
                unrMethCallExpr.typeArgs = unrMethCallExpr.typeArgs.map(x => this.visitType(x));
                unrMethCallExpr.args = unrMethCallExpr.args.map(x => this.visitExpression(x));
            }
            else if (expr is ConditionalExpression condExpr) {
                condExpr.condition = this.visitExpression(condExpr.condition);
                condExpr.whenTrue = this.visitExpression(condExpr.whenTrue);
                condExpr.whenFalse = this.visitExpression(condExpr.whenFalse);
            }
            else if (expr is Identifier ident)
                return this.visitIdentifier(ident);
            else if (expr is UnresolvedNewExpression unrNewExpr) {
                this.visitType(unrNewExpr.cls);
                unrNewExpr.args = unrNewExpr.args.map(x => this.visitExpression(x));
            }
            else if (expr is NewExpression newExpr) {
                this.visitType(newExpr.cls);
                newExpr.args = newExpr.args.map(x => this.visitExpression(x));
            }
            else if (expr is TemplateString templStr)
                return this.visitTemplateString(templStr);
            else if (expr is ParenthesizedExpression parExpr)
                parExpr.expression = this.visitExpression(parExpr.expression);
            else if (expr is UnaryExpression unaryExpr)
                unaryExpr.operand = this.visitExpression(unaryExpr.operand);
            else if (expr is PropertyAccessExpression propAccExpr)
                propAccExpr.object_ = this.visitExpression(propAccExpr.object_);
            else if (expr is ElementAccessExpression elemAccExpr) {
                elemAccExpr.object_ = this.visitExpression(elemAccExpr.object_);
                elemAccExpr.elementExpr = this.visitExpression(elemAccExpr.elementExpr);
            }
            else if (expr is ArrayLiteral arrayLit)
                arrayLit.items = arrayLit.items.map(x => this.visitExpression(x));
            else if (expr is MapLiteral mapLit)
                foreach (var item in mapLit.items)
                    item.value = this.visitExpression(item.value);
            else if (expr is StringLiteral) { }
            else if (expr is BooleanLiteral) { }
            else if (expr is NumericLiteral) { }
            else if (expr is NullLiteral) { }
            else if (expr is RegexLiteral) { }
            else if (expr is CastExpression castExpr) {
                castExpr.newType = this.visitType(castExpr.newType);
                castExpr.expression = this.visitExpression(castExpr.expression);
            }
            else if (expr is InstanceOfExpression instOfExpr) {
                instOfExpr.expr = this.visitExpression(instOfExpr.expr);
                instOfExpr.checkType = this.visitType(instOfExpr.checkType);
            }
            else if (expr is AwaitExpression awaitExpr)
                awaitExpr.expr = this.visitExpression(awaitExpr.expr);
            else if (expr is Lambda lambd)
                return this.visitLambda(lambd);
            else if (expr is ClassReference) { }
            else if (expr is EnumReference) { }
            else if (expr is ThisReference) { }
            else if (expr is StaticThisReference) { }
            else if (expr is MethodParameterReference methParRef)
                return this.visitVariableReference(methParRef);
            else if (expr is VariableDeclarationReference varDeclRef)
                return this.visitVariableReference(varDeclRef);
            else if (expr is ForVariableReference forVarRef)
                return this.visitVariableReference(forVarRef);
            else if (expr is ForeachVariableReference forVarRef2)
                return this.visitVariableReference(forVarRef2);
            else if (expr is CatchVariableReference catchVarRef)
                return this.visitVariableReference(catchVarRef);
            else if (expr is GlobalFunctionReference) { }
            else if (expr is SuperReference) { }
            else if (expr is InstanceFieldReference instFieldRef) {
                instFieldRef.object_ = this.visitExpression(instFieldRef.object_);
                return this.visitVariableReference(instFieldRef);
            }
            else if (expr is InstancePropertyReference instPropRef) {
                instPropRef.object_ = this.visitExpression(instPropRef.object_);
                return this.visitVariableReference(instPropRef);
            }
            else if (expr is StaticFieldReference statFieldRef)
                return this.visitVariableReference(statFieldRef);
            else if (expr is StaticPropertyReference statPropRef)
                return this.visitVariableReference(statPropRef);
            else if (expr is EnumMemberReference) { }
            else if (expr is StaticMethodCallExpression statMethCallExpr) {
                statMethCallExpr.typeArgs = statMethCallExpr.typeArgs.map(x => this.visitType(x));
                statMethCallExpr.args = statMethCallExpr.args.map(x => this.visitExpression(x));
            }
            else if (expr is GlobalFunctionCallExpression globFunctCallExpr)
                globFunctCallExpr.args = globFunctCallExpr.args.map(x => this.visitExpression(x));
            else if (expr is InstanceMethodCallExpression instMethCallExpr) {
                instMethCallExpr.object_ = this.visitExpression(instMethCallExpr.object_);
                instMethCallExpr.typeArgs = instMethCallExpr.typeArgs.map(x => this.visitType(x));
                instMethCallExpr.args = instMethCallExpr.args.map(x => this.visitExpression(x));
            }
            else if (expr is LambdaCallExpression lambdCallExpr)
                lambdCallExpr.args = lambdCallExpr.args.map(x => this.visitExpression(x));
            else
                return this.visitUnknownExpression(expr);
            return expr;
        }
        
        protected virtual void visitMethodParameter(MethodParameter methodParameter)
        {
            this.visitAttributesAndTrivia(methodParameter);
            this.visitVariableWithInitializer(methodParameter);
        }
        
        protected virtual void visitMethodBase(IMethodBase method)
        {
            foreach (var item in method.parameters)
                this.visitMethodParameter(item);
            
            if (method is Constructor const_ && const_.superCallArgs != null)
                for (int i = 0; i < const_.superCallArgs.length(); i++)
                    const_.superCallArgs.set(i, this.visitExpression(const_.superCallArgs.get(i)));
            
            if (method.body != null)
                method.body = this.visitBlock(method.body);
        }
        
        protected void visitMethod(Method method)
        {
            this.currentMethod = method;
            this.currentClosure = method;
            this.visitAttributesAndTrivia(method);
            this.visitMethodBase(method);
            method.returns = this.visitType(method.returns);
            this.currentClosure = null;
            this.currentMethod = null;
        }
        
        protected virtual void visitGlobalFunction(GlobalFunction func)
        {
            this.visitMethodBase(func);
            func.returns = this.visitType(func.returns);
        }
        
        protected void visitConstructor(Constructor constructor)
        {
            this.currentMethod = constructor;
            this.currentClosure = constructor;
            this.visitAttributesAndTrivia(constructor);
            this.visitMethodBase(constructor);
            this.currentClosure = null;
            this.currentMethod = null;
        }
        
        protected virtual void visitField(Field field)
        {
            this.visitAttributesAndTrivia(field);
            this.visitVariableWithInitializer(field);
        }
        
        protected virtual void visitProperty(Property prop)
        {
            this.visitAttributesAndTrivia(prop);
            this.visitVariable(prop);
            if (prop.getter != null)
                prop.getter = this.visitBlock(prop.getter);
            if (prop.setter != null)
                prop.setter = this.visitBlock(prop.setter);
        }
        
        protected virtual void visitInterface(Interface intf)
        {
            this.currentInterface = intf;
            this.visitAttributesAndTrivia(intf);
            intf.baseInterfaces = intf.baseInterfaces.map(x => this.visitType(x));
            foreach (var field in intf.fields)
                this.visitField(field);
            foreach (var method in intf.methods)
                this.visitMethod(method);
            this.currentInterface = null;
        }
        
        protected virtual void visitClass(Class cls)
        {
            this.currentInterface = cls;
            this.visitAttributesAndTrivia(cls);
            if (cls.constructor_ != null)
                this.visitConstructor(cls.constructor_);
            
            cls.baseClass = this.visitType(cls.baseClass);
            cls.baseInterfaces = cls.baseInterfaces.map(x => this.visitType(x));
            foreach (var field in cls.fields)
                this.visitField(field);
            foreach (var prop in cls.properties)
                this.visitProperty(prop);
            foreach (var method in cls.methods)
                this.visitMethod(method);
            this.currentInterface = null;
        }
        
        protected virtual void visitEnum(Enum_ enum_)
        {
            this.visitAttributesAndTrivia(enum_);
            foreach (var value in enum_.values)
                this.visitEnumMember(value);
        }
        
        protected void visitEnumMember(EnumMember enumMember)
        {
            
        }
        
        protected void visitImport(Import imp)
        {
            this.visitAttributesAndTrivia(imp);
        }
        
        public virtual void visitFile(SourceFile sourceFile)
        {
            this.errorMan.resetContext(this);
            this.currentFile = sourceFile;
            foreach (var imp in sourceFile.imports)
                this.visitImport(imp);
            foreach (var enum_ in sourceFile.enums)
                this.visitEnum(enum_);
            foreach (var intf in sourceFile.interfaces)
                this.visitInterface(intf);
            foreach (var cls in sourceFile.classes)
                this.visitClass(cls);
            foreach (var func in sourceFile.funcs)
                this.visitGlobalFunction(func);
            sourceFile.mainBlock = this.visitBlock(sourceFile.mainBlock);
            this.currentFile = null;
        }
        
        public virtual void visitFiles(SourceFile[] files)
        {
            foreach (var file in files)
                this.visitFile(file);
        }
        
        public void visitPackage(Package pkg)
        {
            this.visitFiles(Object.values(pkg.files));
        }
    }
}